tools/ci.sh

```diff
@@ -2,41 +2,153 @@
 ###
 ### Continous Integration Shell Script
 ###
-### This is designed to be the equivalent of what's in .travis.yml, but aimed at executing
+### Designed to be re-entrant on a local dev machine as well, not just on a newly pulled up VM.
 ###
 
-set -euxo pipefail
-
-docker --version
-docker-compose --version
-node --version
-npm --version
-java -version | true
-npm install
-npm run test
-
-cd examples/simple-asset-transfer
-npm install
-rm -rf ./node_modules/websocket/.git/
-rm -rf fabric/api/fabric-client-kv-org*
-cd ./fabric/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run fabric:down
-npm run fabric
-cd ./quorum/api/
-rm -rf ./node_modules/websocket/.git/
-npm install
-cd ../../
-npm run quorum:down
-npm run quorum
-npm run fed:build
-npm run fed:quorum
-docker images
-while docker ps | grep "starting\|unhealthy"; do sleep 15; echo; date; done
-docker ps -a
-npm run fabric:down
-npm run quorum:down
-cd ../..
-#  - npm run test:bc - TODO: fix it, broken now
+startedAt=`date +%s`
+baseDir=`pwd`
+
+function mainTask()
+{
+  set -euxo pipefail
+
+  if ! [ -x "$(command -v lscpu)" ]; then
+    echo 'lscpu is not installed, skipping...'
+  else
+    lscpu
+  fi
+
+  if ! [ -x "$(command -v lsmem)" ]; then
+    echo 'lsmem is not installed, skipping...'
+  else
+    lsmem
+  fi
+
+  docker --version
+  docker-compose --version
+  node --version
+  npm --version
+  java -version || true
+
+  ### COMMON
+
+  rm -rf ./examples/simple-asset-transfer/fabric/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/quorum/api/node_modules
+  rm -rf ./examples/simple-asset-transfer/node_modules
+  rm -rf ./node_modules
+
+  npm install
+  npm run test
+
+  cd examples/simple-asset-transfer
+  npm run fed:quorum:down
+  npm run fabric:down
+  npm run quorum:down
+  npm run quorum:api:down
+
+  npm install
+  rm -rf ./node_modules/websocket/.git/
```

Comment:
do you still need this? same below

Reply:
✅ 32864944bed098a1573df188376095f1067941e8